### DISCUSSION TITLE
How does Fuse.js search work? Cannot search newly added file

### DISCUSSION THREAD

**Discussion by FuadEfendi** [razonyang/hugo-theme-bootstrap#716] [Q&A] [answered]
Hello, 

I have few thousand files already, added new file and can't search it even after several restarts; how does it work? Should remove some temporary files? Should I wait after restart?
I added specific keyword to this file in few places; search returns few other documents but not this one. 

Thanks,

**Comment by razonyang** [comment, marked as answer]
In short.

1. The theme generates the `index.json` which contains all pages, i.e. https://hbs.razonyang.com/v1/v1/en/search/index.json.
2. The search page loads the `index.json` **once** per page load, parse the user input, then display the search result.

So, if you're adding new pages, you should reload the search page for reloading the index.json.
If it's still not working, please check whether your new pages are included in `index.json`.

> Force refresh `CTRL+F5` and restrart `hugo server` may ensure that is no cache.

**Comment by FuadEfendi** [comment]
Thank you for confirming Razon Yang. Yes, I suspected caching, it doesn't happen in production Netlify possibly because I use "must revalidate" HTTP headers. Only locally.

My `index.json` is more than 1Mb already, and download takes 37 seconds WOW ;) time to migrate to something else although I love Fuse.js approach... for example, I have special term with letter "K" in it, such as "K Theory", and all my attempts to tune search fail. Needless to mention index size will quickly become over 10Mb...

**Reply by FuadEfendi** [reply]
although comparable to "single page" (Angular) approach, 10Mb index size is Ok.

This explains issue I experienced: after restarting server, do a search (which won't return expected results); and after 1-2 minutes do search again, and results will be as expected. Async download of JSON index.

**Reply by razonyang** [reply]
You're welcome.

Yep, this is a limitation of local search, the search function need all the pages' info, there will be a problem if the index files is too large.

You can enable compression on server side to make `index.json` transfer/download faster, such as `gzip`.

The docs site's(about 1k pages) index.json size:

| Uncompressed | Compressed |
|---|---|
| `401kB` | `67.5kB` |

If your site is too large(over 10k pages), it is recommended to use external search services, which can filter the results on server side.
